packages/xsnap/src/xsnap.c

```diff
@@ -0,0 +1,1350 @@
+		1 * 1024 * 1024,	/* incrementalHeapCount */
+		4096,				/* stackCount */
+		32000,				/* keyCount */
+		1993,				/* nameModulo */
+		127,				/* symbolModulo */
+		8192 * 1024,		/* parserBufferSize */
+		1993,				/* parserTableModulo */
+	};
+	xsCreation* creation = &_creation;
+
+	txSnapshot snapshot = {
+		"xsnap 0.1.0",
+		11,
+		gxSnapshotCallbacks,
+		mxSnapshotCallbackCount,
+		fxSnapshotRead,
+		fxSnapshotWrite,
+		NULL,
+		0,
+		NULL,
+		NULL,
+		NULL,
+	};
+
+	xsMachine* machine;
+	char *path;
+	char* dot;
+
+	for (argi = 1; argi < argc; argi++) {
+		if (argv[argi][0] != '-')
+			continue;
+		if (!strcmp(argv[argi], "-f")) {
+			freeze = 1;
+		}
+		else if (!strcmp(argv[argi], "-h")) {
+			fxPrintUsage();
+			return 0;
+		} else if (!strcmp(argv[argi], "-i")) {
+			argi++;
+			if (argi < argc)
+				interval = atoi(argv[argi]);
+			else {
+				fxPrintUsage();
+				return 1;
+			}
+		}
+		else if (!strcmp(argv[argi], "-l")) {
+			argi++;
+			if (argi < argc)
+				gxMeteringLimit = atoi(argv[argi]);
+			else {
+				fxPrintUsage();
+				return 1;
+			}
+		}
+		else if (!strcmp(argv[argi], "-p"))
+			gxMeteringPrint = 1;
+		else if (!strcmp(argv[argi], "-r")) {
+			argi++;
+			if (argi < argc)
+				argr = argi;
+			else {
+				fxPrintUsage();
+				return 1;
+			}
+		}
+		else if (!strcmp(argv[argi], "-v")) {
+			printf("XS %d.%d.%d\n", XS_MAJOR_VERSION, XS_MINOR_VERSION, XS_PATCH_VERSION);
+			return 0;
+		} else {
+			fxPrintUsage();
+			return 1;
+		}
+	}
+	if (gxMeteringLimit) {
+		if (interval == 0)
+			interval = 1;
+	}
+	fxInitializeSharedCluster();
+	if (argr) {
+		snapshot.stream = fopen(argv[argr], "rb");
+		if (snapshot.stream) {
+			machine = fxReadSnapshot(&snapshot, "xsnap", NULL);
+			fclose(snapshot.stream);
+		}
+		else
+			snapshot.error = errno;
+		if (snapshot.error) {
+			fprintf(stderr, "cannot read snapshot %s: %s\n", argv[argr], strerror(snapshot.error));
+			return 1;
+		}
+	}
+	else {
+		machine = xsCreateMachine(creation, "xsnap", NULL);
+		fxBuildAgent(machine);
+		fxPatchBuiltIns(machine);
+	}
+	if (freeze) {
+		fxFreezeBuiltIns(machine);
+		fxShareMachine(machine);
+		fxCheckAliases(machine);
+		machine = xsCloneMachine(creation, machine, "xsnap", NULL);
+	}
+	if (!(fromParent = fdopen(3, "rb"))) {
+		fprintf(stderr, "fdopen(3) from parent failed\n");
+		c_exit(1);
+	}
+	if (!(toParent = fdopen(4, "wb"))) {
+		fprintf(stderr, "fdopen(4) to parent failed\n");
+		c_exit(1);
+	}
+	xsBeginMetering(machine, fxMeteringCallback, interval);
+	{
+		char done = 0;
+		while (!done) {
+			char* nsbuf;
+			size_t nslen;
+			int readError = fxReadNetString(fromParent, &nsbuf, &nslen);
+			if (readError != 0) {
+				if (feof(fromParent)) {
+					break;
+				} else {
+					fprintf(stderr, "%s\n", fxReadNetStringError(readError));
+					c_exit(1);
+				}
+			}
+			char command = *nsbuf;
+			// fprintf(stderr, "command: len %d %c arg: %s\n", nslen, command, nsbuf + 1);
+			switch(command) {
+			case '?':
```

Comment:
For symmetry with outbound syscalls, I changed the 'd' code to '?'.

Reply:
I don't really see the symmetry. I wonder about `>` for "send". But I don't feel strongly.